src/SquidMath.h

```diff
@@ -123,5 +123,53 @@ SetToNaturalSumOrMax(S &var, Args... args)
     return var;
 }
 
+// If NaturalProduct() performance becomes important, consider using GCC and clang
+// built-ins like __builtin_mul_overflow() instead of manual overflow checks.
+
+/// \returns an exact, non-overflowing product of the arguments (or nothing)
+/// \returns nothing if at least one of the arguments is negative
+template <typename P, typename T, typename U>
+Optional<P>
+NaturalProduct(T t, U u)
+{
+    // ensure that the shifting below will work
+    static_assert(std::is_integral<T>::value, "the first argument is integral");
+    static_assert(std::is_integral<U>::value, "the second argument is integral");
+
+    // assume that callers treat negative numbers specially (see IncreaseSum() for details)
+    if (Less(t, 0) || Less(u, 0))
+        return Optional<P>();
+
+    // check that both operands do not overflow the result type
+    if (!IncreaseSum(P(0), t).has_value() || !IncreaseSum(P(0), u).has_value())
+        return Optional<P>();
+
+    auto p = static_cast<P>(t);
+    auto result = Optional<P>(0);
+
+    while (u) {
+        if (u & 1) {
+            result = IncreaseSum(result.value(), p);
+            if (!result.has_value())
+                return Optional<P>();
+        }
+        u >>= 1;
+        if (!u)
+            break;
+        if (!IncreaseSum(p, p).has_value())
+            return Optional<P>();
+        p <<= 1;
+    }
```

Comment:
I have mentioned the problem of detecting integer multiplication overflow to another developer, and they suggested a very basic algorithm: Given non-negative integers, `a*b` overflows for type `A` (i.e. `a*b > maxA)` if and only if `a` is not zero and `b > (maxA/a)`, where the division is regular integer division. I do not see anything wrong with that algorithm. Do you? It is essentially the same algorithm we use for integer sum overflow detection... I know it needs to be carefully adjusted to use a third type `P` instead of `A`, but that feels doable. If it works, please replace the loop-based algorithm with this division-based one.